### PR TITLE
Keep track of ProjectRootElements that change via save to a stream Fixes #4824

### LINKED ISSUE #4824: Save in ClickOnce scenario does not update ProjectRootElement._lastWriteTimeWhenRead (regression)
See [AB#984122](https://devdiv.visualstudio.com/0bdbc590-a062-4c3f-b0f6-9383f67865ee/_workitems/edit/984122)

More Info: 
1. The bug reproduces on master-29312.03 & Dev16.3 Pre4 d16.3-29311.1183.
2. This bug doesn't reproduce on master-29311.112 & Dev16.3 Pre3 d16.3-29230.61.
3. This bug doesn't repro when setting NuGet package management format to Packages.config.
4. This bug doesn't repro after rebuilt the project. 

Steps to reproduce:
1. Create a project, such as WPF/Winforms project.
2. Right click on Reference open Manage NuGet Packages… 
3. Click Settings and set the package management format to PackageReference .
4. Install a NuGet package such as Newtonsoft.Json.
5. Publish and install the project.
6. On Signing page uncheck the 'Sign the ClickOnce manifests'.
7. Publish and install the project.

Actual result: 
Cannot start the application

The warning information is:
Application validation did not succeed. Unable to continue.
See Details txt file in attachment.
Expect result:
The application should be add successfully.

### PR DESCRIPTION
**Pull request by Forgind**
Writing to a stream may or may not involve modifying a project, so we previously did not update _lastWriteTimeWhenRead to reflect changes from a stream. When we changed how we calculate MSBuildAllProjects to only include the most recently modified project (so we don't have to look through a potentially long list of irrelevant projects), we started missing out on project files that were written via writing a stream but not modified explicitly with a non-stream write—these would not be added to MSBuildAllProjects because they weren't the "most recently modified" project, causing them never to be built if the most recently written project according to MSBuildAllProjects had been written before the last build. This caused the error in ClickOnce publishing, so this Fixes #4824.

### CODE REVIEW

File: ref/Microsoft.Build/net/Microsoft.Build.cs [hunk 1]
```diff
@@ -226,6 +226,7 @@ public partial class ProjectPropertyGroupElement : Microsoft.Build.Construction.
     public partial class ProjectRootElement : Microsoft.Build.Construction.ProjectElementContainer
     {
         internal ProjectRootElement() { }
+        public System.Nullable<System.DateTime> StreamTime;
```
Comment (rainersigwald) [inline ref/Microsoft.Build/net/Microsoft.Build.cs]: Does this actually need to be public, or can it be `internal`?

File: src/Build/Evaluation/Evaluator.cs [hunk 2]
```diff
@@ -1988,6 +1993,11 @@ private List<ProjectRootElement> ExpandAndLoadImports(string directoryOfImportin
                                 _lastModifiedProject = importedProjectElement;
                             }
 
+                            if (importedProjectElement.StreamTime != null && importedProjectElement.StreamTime > _lastModifiedProject.LastWriteTimeWhenRead)
+                            {
+                                _streamImports.Append(";").Append(importedProjectElement.FullPath);
```
Comment (rainersigwald) [inline src/Build/Evaluation/Evaluator.cs]: I think it would be less allocatey to keep a list of elements and do a `Join(';')` at the end.

File: src/Build/Construction/ProjectRootElement.cs [hunk 3]
```diff
@@ -1584,6 +1586,7 @@ public void Save(TextWriter writer)
                 XmlDocument.Save(projectWriter);
             }
 
+            StreamTime = DateTime.Now;
```
Comment (jeffkl) [inline src/Build/Construction/ProjectRootElement.cs]: Is this guaranteed to write to disk and to the same path that represents the file?  Why not just update the last write time?  If its a memory stream, or anything not on disk, then it wouldn't trigger a rebuild anyway since the datetime stamp would be calculated off of the file during build.  So it seems like it would be better to just assume the file was updated on disk if the stream was updated so everything works the same.  Unless of course I'm missing something.

Comment (Forgind) [reply src/Build/Construction/ProjectRootElement.cs]: The "last modified project" [is calculated](https://github.com/microsoft/msbuild/blob/master/src/Build/Evaluation/Evaluator.cs#L1986-L1989) purely based on `_lastWriteTimeWhenRead`. Since we only keep the last modified project to add to MSBuildAllProjects, it would break that should the last thing we wrote to not be a file on disk.

Comment (jeffkl) [reply src/Build/Construction/ProjectRootElement.cs]: > Since we only keep the last modified project to add to MSBuildAllProjects, it would break that should the last thing we wrote to not be a file on disk.

I'm still confused.  If we didn't write a file to disk, the up-to-date check wouldn't pick it up either right?  Just adding the path of the import/project to MSBuildAllProjects isn't enough, the timestamp on disk has to change as well.

Maybe we just need a new LastWriteTime property which reflects when the file is modified but doesn't cause the file to be re-read from disk?  And then this calculation would come from that.

Comment (Forgind) [reply src/Build/Construction/ProjectRootElement.cs]: This is the scenario I mean:
```
Write to file a
Build
Write to file b
Write to stream c that is not represented by a file on disk
Build
```
In this case, if we update the last write time of the projectrootelement corresponding to c, then in the part of Evaluator I linked above, it will think the last modified project was c, not b, so it will prepend that to MSBuildAllProjects.
When we check to see if we need to rebuild, there's no corresponding file on disk for the last project "written to," so since we only kept the last project (for performance) instead of keeping all of them, our up-to-date check fails—we miss that file b was written to.
In other words, the problem is that when we write to a stream, it could be writing to a file (in which case, we should update _lastWriteTimeWhenRead) or not, in which case doing so would prevent us from noticing the real most recently written ProjectRootElement.

Could you explain your LastWriteTime idea a little more and how we can detect that a file has been modified without reading it? I don't understand how it differs from _lastWriteTimeWhenRead.

File: src/Build/Construction/ProjectRootElement.cs [hunk 4]
```diff
@@ -619,6 +619,8 @@ public int Version
         /// </summary>
         public DateTime LastWriteTimeWhenRead => Link != null ? RootLink.LastWriteTimeWhenRead : _lastWriteTimeWhenRead;
 
+        internal DateTime? StreamTime = null;
```
Comment (danmosemsft) [inline src/Build/Construction/ProjectRootElement.cs]: Seems that some DateTime's in this type are stored as Utc and some are not. Would it be helpful to consistently use UTC? Or else maybe consider a suffix on the field name eg `NotUtc`

Comment (Forgind) [reply src/Build/Construction/ProjectRootElement.cs]: It's important for StreamTime to be in the same time zone as the file's time and _lastWriteTimeWhenRead, and I think using local time is the least confusing. I changed it to explicitly convert it to local time just for clarity.

Comment (rainersigwald) [reply src/Build/Construction/ProjectRootElement.cs]: I prefer the other direction: use UTC everywhere. That's what we do for most things. Unfortunately, we didn't catch this when we added `LastWriteTimeWhenRead` as a public field instead of `LastWriteTimeWhenReadUtc`. So I guess converging on local time in this class is probably the way to go.

File: src/Build/Evaluation/Evaluator.cs [hunk 5]
```diff
@@ -2367,12 +2378,12 @@ private void SetAllProjectsProperty()
...
                 P oldValue = _data.GetProperty(Constants.MSBuildAllProjectsPropertyName);
-
+                string streamImports = string.Join(";", _streamImports.ToArray());
                 P newValue = _data.SetProperty(
                     Constants.MSBuildAllProjectsPropertyName,
                     oldValue == null
-                        ? _lastModifiedProject.FullPath
-                        : $"{_lastModifiedProject.FullPath};{oldValue.EvaluatedValue}",
+                        ? $"{_lastModifiedProject.FullPath};{streamImports}"
+                        : $"{_lastModifiedProject.FullPath};{streamImports};{oldValue.EvaluatedValue}",
```
Comment (rainersigwald) [inline src/Build/Evaluation/Evaluator.cs]: Hmm. I don't love having extra semicolons in this list, which we will now in the "normal" (no stream save) case. But not a huge deal, everything should be perfectly capable of stripping them out later.

File: src/Build/Construction/ProjectRootElement.cs (hunk 4 above)
Comment (danmosemsft) [reply src/Build/Construction/ProjectRootElement.cs]: The public properties need not have the Utc suffix, but the backing fields could be Utc, and have utc suffix, without breaking anyone (unless the machine is changing timezones)

File: src/Build/Construction/ProjectRootElement.cs [hunk 6]
```diff
@@ -1584,6 +1590,7 @@ public void Save(TextWriter writer)
                 XmlDocument.Save(projectWriter);
             }
 
+            StreamTime = DateTime.Now.ToUniversalTime();
```
Comment (rainersigwald) [inline src/Build/Construction/ProjectRootElement.cs]:
```suggestion
            StreamTime = DateTime.UtcNow;
```

File: src/Build/Evaluation/Evaluator.cs [hunk 7]
```diff
@@ -1988,6 +1995,12 @@ private List<ProjectRootElement> ExpandAndLoadImports(string directoryOfImportin
                                 _lastModifiedProject = importedProjectElement;
                             }
 
+                            if (importedProjectElement.StreamTime != null && importedProjectElement.StreamTime?.ToLocalTime() > _lastModifiedProject.LastWriteTimeWhenRead)
```
[the remark this answers is not in the document]
Comment (Forgind) [reply src/Build/Evaluation/Evaluator.cs]: DateTime? doesn't have ToLocalTime() without the ?, so I removed the first null check instead.

[the remark this answers is not in the document]
Comment (rainersigwald) [reply src/Build/Evaluation/Evaluator.cs]: Ah, you can do `.Value` to get through a nullable. But removing the first one sounds good.

File: src/Build/Construction/ProjectRootElement.cs [hunk 8]
```diff
@@ -617,7 +617,7 @@ public int Version
         /// This can be used to see whether the file has been changed on disk
         /// by an external means.
         /// </summary>
-        public DateTime LastWriteTimeWhenRead => Link != null ? RootLink.LastWriteTimeWhenRead : _lastWriteTimeWhenRead.ToLocalTime();
+        public DateTime LastWriteTimeWhenRead => Link != null ? RootLink.LastWriteTimeWhenRead : _lastWriteTimeWhenReadUtc.ToLocalTime();
 
         internal DateTime? StreamTime = null;
```
Comment (rainersigwald) [inline src/Build/Construction/ProjectRootElement.cs]: Should this be 
```suggestion
        internal DateTime? StreamTimeUtc = null;
```

?
